feat: Writed Unit and integration test

rmnkubyshkin opened the pull request:
feat: Writed Unit and integration test
	1. Unit test for all slices, root reducer
	2. Integration test for constructor, modal view, mock api request, ingredients

rmnkubyshkin committed:
fix: Added some functional and selectors to slices
	1. Moved each slice to individual folder
	2. Added new reducers and selectors for some slices
	3. Edited App architecture, after integration tests.
	All routers moved to single Routes. Modal viewies was moved out
	from main Routes.
	4. Small editions on other components
feat: Writed Unit and integration test
	1. Unit test for all slices, root reducer
	2. Integration test for constructor, modal view, mock api
	   request, ingredients

aromanovnik commented on cypress/e2e/general/general.cy.ts:
```diff
@@ -0,0 +1,115 @@
+
+    cy.intercept('GET', '**/auth/user', { fixture: 'user.json' }).as('getUser');
+
+    cy.visit('http://192.168.11.138:4000/');
```
**Можно лучше**: По умолчанию лучше использовать `localhost:4000`, чтобы тесты не зависели от конкретного IP-адреса. Если требуется нестандартный адрес, его лучше вынести в константу или использовать переменные окружения для удобства настройки.

aromanovnik commented on cypress/e2e/general/general.cy.ts:
```diff
@@ -0,0 +1,115 @@
+  });
+
+  after(() => {
+    cy.window().then((win) => {
```
**Можно лучше**: Вместо явного удаления каждого элемента локального хранилища и куков лучше использовать `cy.clearLocalStorage();` и `cy.clearCookies();`, чтобы очистить их полностью и избежать потенциальных ошибок, если в будущем добавятся новые ключи.

aromanovnik commented on cypress/e2e/general/general.cy.ts:
```diff
@@ -0,0 +1,115 @@
+import login from '../../fixtures/login.json';
+import orders from '../../fixtures/orders.json';
+describe('Тестирование общее', () => {
```
**Надо исправить**: По заданию должны быть тесты на открытие и закрытие модального окна с описанием ингредиента, а также проверка, что в модальном окне отображаются данные именно того ингредиента, по которому был клик. Сейчас этих тестов нет, их необходимо добавить.

<img width="721" alt="Снимок экрана 2025-03-17 в 23 09 32" src="https://github.com/user-attachments/assets/7439ad72-0836-4103-9950-05215bc156ab" />

aromanovnik commented on cypress/e2e/general/general.cy.ts:
```diff
@@ -0,0 +1,115 @@
+    it('Должен создать заказ', () => {
+      cy.wait('@getIngredients');
+      cy.contains('Краторная булка N-200i').should('exist');
+      cy.contains('Добавить').should('exist').click();
```
**Надо исправить**: Перед добавлением ингредиента важно проверить, что его нет в конструкторе. В текущей реализации проверяется только наличие ингредиента, но не факт его добавления через конкретное действие (например, нажатие кнопки “Добавить”). Для корректного тестирования любого действия необходимо:
1️⃣ Убедиться, что элемента нет.
2️⃣ Выполнить действие.
3️⃣ Проверить, что элемент появился в разметке.
Этот подход стоит применить ко всем проверкам. ✅

aromanovnik commented on cypress/e2e/general/general.cy.ts:
```diff
@@ -0,0 +1,115 @@
+    });
+  });
+  describe('Перехватывает запрос на создание заказа', () => {
+    it('Должен перехватывать запрос на создание заказа', () => {
```
**Можно лучше**: Стоит добавить проверку, что кнопка “Оформить заказ” активируется только после добавления булки и хотя бы одного ингредиента. Это поможет убедиться, что функциональность реализована корректно.

aromanovnik commented on src/services/slices/cookingSlice/cookingSlice.test.ts:
```diff
@@ -0,0 +1,307 @@
+  selectCookingItemsLoading,
+  setCookingItems
+} from './cookingSlice';
+import { error } from 'console';
```
**Можно лучше**: Неиспользованные импорты, методы и функции и т.д. лучше удалять из проекта, так как они утяжеляют кодовую базу, создают визуальный шум и могут сбивать с толку разработчиков, читающих код. Это также улучшает производительность и облегчает поддержку проекта, снижая вероятность ошибок.

aromanovnik commented on src/services/slices/cookingSlice/cookingSlice.test.ts:
```diff
@@ -0,0 +1,307 @@
+  image_large: 'https://code.s3.yandex.net/react/code/meat-01-large.png'
+};
+
+describe('Тесты для cookingSlice', () => {
```
**Можно лучше**: Стоит добавить тесты для крайних ситуаций, таких как удаление ингредиента из пустого конструктора и перемещение ингредиента с неверными индексами. Это поможет убедиться, что логика конструктора работает корректно в различных нестандартных ситуациях.

aromanovnik commented on src/services/slices/userSlice/userSlice.test.ts:
```diff
@@ -0,0 +1,542 @@
+  refreshToken: 'refreshToken'
+};
+
+describe('Тесты для userSlice', () => {
```
**Надо исправить**: По заданию должны быть тесты, проверяющие, как редьюсер обрабатывает экшены, создаваемые `createAsyncThunk`. Необходимо добавить проверки для экшенов `.pending`, `.fulfilled` и `.rejected`, чтобы убедиться, что состояние редьюсера корректно изменяется при разных исходах запроса.

<img width="776" alt="Снимок экрана 2025-03-17 в 23 34 31" src="https://github.com/user-attachments/assets/880a328e-7106-434d-b951-6f1f922fe261" />

aromanovnik reviewed (changes requested):
Добрый день, Roman! 👋 *(Нужно развернуть общий комментарий ↓)*

Работа выполнена на высоком уровне! В проект добавлены и настроены библиотеки Cypress и Jest, а команды для их запуска описаны в разделе `scripts` файла package.json. 🔧 Тесты структурированы в блоки `describe` с понятным описанием проверяемой функциональности, а внутри каждого теста чётко указано, что именно проверяется. ✅ Набор проверок полностью охватывает тестируемую функциональность, и всё это написано на TypeScript. Все тесты успешно завершаются. Отличная работа! 🚀

Тем не менее, есть несколько замечаний, которые нужно доработать. 🔧 Более детальная информация указана в комментариях к коду. 💡

**Что нужно исправить**:
* По заданию должны быть тесты на открытие и закрытие модального окна с описанием ингредиента, а также проверка, что в модальном окне отображаются данные именно того ингредиента, по которому был клик. Сейчас этих тестов нет, их необходимо добавить.
* По заданию должны быть тесты, проверяющие, как редьюсер обрабатывает экшены, создаваемые `createAsyncThunk`. Необходимо добавить проверки для экшенов `.pending`, `.fulfilled` и `.rejected`, чтобы убедиться, что состояние редьюсера корректно изменяется при разных исходах запроса.
* Перед добавлением ингредиента важно проверить, что его нет в конструкторе. В текущей реализации проверяется только наличие ингредиента, но не факт его добавления через конкретное действие (например, нажатие кнопки “Добавить”). Для корректного тестирования любого действия необходимо:
1️⃣ Убедиться, что элемента нет.
2️⃣ Выполнить действие.
3️⃣ Проверить, что элемент появился в разметке.
Этот подход стоит применить ко всем проверкам. ✅

**Что можно улучшить**:
* В репозиторий лучше добавлять файл `.env.example` с пустыми или случайными значениями. Это позволит другим разработчикам понять, какие переменные окружения требуются, без риска случайного раскрытия конфиденциальной информации. 
* По умолчанию лучше использовать `localhost:4000`, чтобы тесты не зависели от конкретного IP-адреса. Если требуется нестандартный адрес, его лучше вынести в константу или использовать переменные окружения для удобства настройки.
* Вместо явного удаления каждого элемента локального хранилища и куков лучше использовать `cy.clearLocalStorage();` и `cy.clearCookies();`, чтобы очистить их полностью и избежать потенциальных ошибок, если в будущем добавятся новые ключи.
* Стоит добавить проверку, что кнопка “Оформить заказ” активируется только после добавления булки и хотя бы одного ингредиента. Это поможет убедиться, что функциональность реализована корректно.
* Неиспользованные импорты, методы и функции и т.д. лучше удалять из проекта, так как они утяжеляют кодовую базу, создают визуальный шум и могут сбивать с толку разработчиков, читающих код. Это также улучшает производительность и облегчает поддержку проекта, снижая вероятность ошибок.
* Стоит добавить тесты для крайних ситуаций, таких как удаление ингредиента из пустого конструктора и перемещение ингредиента с неверными индексами. Это поможет убедиться, что логика конструктора работает корректно в различных нестандартных ситуациях.

Напоминаю, что работа может быть принята только после исправления всех критических замечаний **Надо исправить**.

Удачного рефакторинга кода.

rmnkubyshkin committed:
fix: Corrected unit and integration tests

	1. Added check open/close modal view by ingredient click
	2. Added check start state in each integration test
	3. Added on each unit test with async reducers check three state
	   as pending/rejected/fullfiled

aromanovnik reviewed (approved):
🎉 **Поздравляю! Работа принята!**

Все недочёты успешно исправлены, и результат получился просто отличным! 🌟

💪 Вы проделали огромную работу, и это действительно впечатляет!
🚀 Желаю успехов в дальнейшей учёбе и новых достижений!
